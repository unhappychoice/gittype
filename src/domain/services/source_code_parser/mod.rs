mod cache_builder;
pub mod chunk_extractor;
mod comment_processor;
mod indent_processor;
pub mod parsers;
pub mod source_code_parser;

pub use cache_builder::CacheBuilder;
pub use chunk_extractor::{ChunkExtractor, ParentChunk};
pub use comment_processor::CommentProcessor;
pub use indent_processor::IndentProcessor;
pub use source_code_parser::SourceCodeParser;
