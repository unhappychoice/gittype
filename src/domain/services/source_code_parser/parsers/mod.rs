use crate::domain::models::languages::{
    CSharp, Clojure, Cpp, Dart, Go, Haskell, Java, JavaScript, Kotlin, Php, Python, Ruby, Rust,
    Scala, Swift, TypeScript, Zig, C,
};
use crate::domain::models::ChunkType;
use crate::domain::models::Language;
use crate::{GitTypeError, Result};
use once_cell::sync::Lazy;
use std::cell::RefCell;
use std::collections::HashMap;
use tree_sitter::{Node, Parser, Query, Tree};

pub mod c;
pub mod clojure;
pub mod cpp;
pub mod csharp;
pub mod dart;
pub mod go;
pub mod haskell;
pub mod java;
pub mod javascript;
pub mod kotlin;
pub mod php;
pub mod python;
pub mod ruby;
pub mod rust;
pub mod scala;
pub mod swift;
pub mod typescript;
pub mod zig;

pub trait LanguageExtractor {
    fn tree_sitter_language(&self) -> tree_sitter::Language;

    fn comment_query(&self) -> &str;

    fn query_patterns(&self) -> &str;
    fn capture_name_to_chunk_type(&self, capture_name: &str) -> Option<ChunkType>;
    fn extract_name(&self, node: Node, source_code: &str, capture_name: &str) -> Option<String>;

    fn middle_implementation_query(&self) -> &str;
    fn middle_capture_name_to_chunk_type(&self, _capture_name: &str) -> Option<ChunkType>;
}

type ParserFactory = fn() -> Result<Parser>;
type ExtractorFactory = fn() -> Box<dyn LanguageExtractor>;

pub struct ParserRegistry {
    parsers: HashMap<String, ParserFactory>,
    extractors: HashMap<String, ExtractorFactory>,
}

impl ParserRegistry {
    fn new() -> Self {
        let mut registry = Self {
            parsers: HashMap::new(),
            extractors: HashMap::new(),
        };

        // Register all supported languages using a macro to reduce repetition
        macro_rules! register_language {
            ($lang:ident, $module:ident, $extractor:ident) => {
                registry.register(
                    $lang.name().to_string(),
                    $module::$extractor::create_parser,
                    || Box::new($module::$extractor),
                );
            };
        }

        register_language!(C, c, CExtractor);
        register_language!(Clojure, clojure, ClojureExtractor);
        register_language!(Cpp, cpp, CppExtractor);
        register_language!(CSharp, csharp, CSharpExtractor);
        register_language!(Dart, dart, DartExtractor);
        register_language!(Go, go, GoExtractor);
        register_language!(Haskell, haskell, HaskellExtractor);
        register_language!(Java, java, JavaExtractor);
        register_language!(JavaScript, javascript, JavaScriptExtractor);
        register_language!(Kotlin, kotlin, KotlinExtractor);
        register_language!(Php, php, PhpExtractor);
        register_language!(Python, python, PythonExtractor);
        register_language!(TypeScript, typescript, TypeScriptExtractor);
        register_language!(Ruby, ruby, RubyExtractor);
        register_language!(Rust, rust, RustExtractor);
        register_language!(Scala, scala, ScalaExtractor);
        register_language!(Swift, swift, SwiftExtractor);
        register_language!(Zig, zig, ZigExtractor);

        registry
    }

    fn register(
        &mut self,
        language: String,
        parser_factory: ParserFactory,
        extractor_factory: ExtractorFactory,
    ) {
        self.parsers.insert(language.clone(), parser_factory);
        self.extractors.insert(language, extractor_factory);
    }

    pub fn create_parser(&self, language: &str) -> Result<Parser> {
        self.parsers
            .get(language)
            .ok_or_else(|| {
                GitTypeError::ExtractionFailed(format!("Unsupported language: {}", language))
            })
            .and_then(|factory| factory())
    }

    pub fn get_extractor(&self, language: &str) -> Result<Box<dyn LanguageExtractor>> {
        self.extractors
            .get(language)
            .ok_or_else(|| {
                GitTypeError::ExtractionFailed(format!("Unsupported language: {}", language))
            })
            .map(|factory| factory())
    }

    pub fn create_query(&self, language: &str) -> Result<Query> {
        let extractor = self.get_extractor(language)?;
        let tree_sitter_lang = extractor.tree_sitter_language();
        let query_str = extractor.query_patterns();

        Query::new(&tree_sitter_lang, query_str).map_err(|e| {
            GitTypeError::ExtractionFailed(format!(
                "Failed to create query for {}: {}",
                language, e
            ))
        })
    }

    pub fn create_comment_query(&self, language: &str) -> Result<Query> {
        let extractor = self.get_extractor(language)?;
        let tree_sitter_lang = extractor.tree_sitter_language();
        let query_str = extractor.comment_query();

        Query::new(&tree_sitter_lang, query_str).map_err(|e| {
            GitTypeError::ExtractionFailed(format!(
                "Failed to create comment query for {}: {}",
                language, e
            ))
        })
    }

    pub fn create_middle_implementation_query(&self, language: &str) -> Result<Query> {
        let extractor = self.get_extractor(language)?;
        let tree_sitter_lang = extractor.tree_sitter_language();
        let query_str = extractor.middle_implementation_query();

        // If query is empty, create a dummy query that matches nothing
        let actual_query = if query_str.trim().is_empty() {
            "(ERROR) @dummy" // This will never match anything but is syntactically valid
        } else {
            query_str
        };

        Query::new(&tree_sitter_lang, actual_query).map_err(|e| {
            GitTypeError::ExtractionFailed(format!(
                "Failed to create middle implementation query for {}: {}",
                language, e
            ))
        })
    }

    pub fn supported_languages(&self) -> Vec<String> {
        self.parsers.keys().cloned().collect()
    }
}

static REGISTRY: Lazy<ParserRegistry> = Lazy::new(ParserRegistry::new);

pub fn get_parser_registry() -> &'static ParserRegistry {
    &REGISTRY
}

thread_local! {
    static TL_PARSERS: RefCell<HashMap<String, Parser>> = RefCell::new(HashMap::new());
}

/// Parse source using a thread-local parser per language to avoid re-allocations.
pub fn parse_with_thread_local(language: &str, content: &str) -> Option<Tree> {
    TL_PARSERS.with(|cell| {
        let mut map = cell.borrow_mut();
        let parser = match map.get_mut(language) {
            Some(p) => p,
            None => {
                // Create and insert parser if not exists
                match REGISTRY.create_parser(language) {
                    Ok(p) => {
                        map.insert(language.to_string(), p);
                        map.get_mut(language).unwrap()
                    }
                    Err(_) => return None,
                }
            }
        };
        parser.parse(content, None)
    })
}
