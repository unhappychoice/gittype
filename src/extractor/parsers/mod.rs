use crate::extractor::models::{
    language::{
        CSharp, Cpp, Dart, Go, Haskell, Java, JavaScript, Kotlin, Language, Php, Python, Ruby,
        Rust, Swift, TypeScript, C,
    },
};
use crate::models::ChunkType;
use crate::{GitTypeError, Result};
use once_cell::sync::Lazy;
use std::cell::RefCell;
use std::collections::HashMap;
use tree_sitter::{Node, Parser, Query, Tree};

pub mod c;
pub mod cpp;
pub mod csharp;
pub mod dart;
pub mod go;
pub mod haskell;
pub mod java;
pub mod javascript;
pub mod kotlin;
pub mod php;
pub mod python;
pub mod ruby;
pub mod rust;
pub mod swift;
pub mod typescript;

pub trait LanguageExtractor {
    fn tree_sitter_language(&self) -> tree_sitter::Language;
    fn query_patterns(&self) -> &str;
    fn comment_query(&self) -> &str;
    fn capture_name_to_chunk_type(&self, capture_name: &str) -> Option<ChunkType>;
    fn extract_name(&self, node: Node, source_code: &str, capture_name: &str) -> Option<String>;
}

type ParserFactory = fn() -> Result<Parser>;
type ExtractorFactory = fn() -> Box<dyn LanguageExtractor>;

pub struct ParserRegistry {
    parsers: HashMap<String, ParserFactory>,
    extractors: HashMap<String, ExtractorFactory>,
}

impl ParserRegistry {
    fn new() -> Self {
        let mut registry = Self {
            parsers: HashMap::new(),
            extractors: HashMap::new(),
        };

        // Register all supported languages
        registry.register(
            Rust.name().to_string(),
            rust::RustExtractor::create_parser,
            || Box::new(rust::RustExtractor),
        );

        registry.register(
            TypeScript.name().to_string(),
            typescript::TypeScriptExtractor::create_parser,
            || Box::new(typescript::TypeScriptExtractor),
        );

        registry.register(
            JavaScript.name().to_string(),
            javascript::JavaScriptExtractor::create_parser,
            || Box::new(javascript::JavaScriptExtractor),
        );

        registry.register(
            Python.name().to_string(),
            python::PythonExtractor::create_parser,
            || Box::new(python::PythonExtractor),
        );

        registry.register(
            Ruby.name().to_string(),
            ruby::RubyExtractor::create_parser,
            || Box::new(ruby::RubyExtractor),
        );

        registry.register(
            Go.name().to_string(),
            go::GoExtractor::create_parser,
            || Box::new(go::GoExtractor),
        );

        registry.register(
            Swift.name().to_string(),
            swift::SwiftExtractor::create_parser,
            || Box::new(swift::SwiftExtractor),
        );

        registry.register(
            Kotlin.name().to_string(),
            kotlin::KotlinExtractor::create_parser,
            || Box::new(kotlin::KotlinExtractor),
        );

        registry.register(
            Java.name().to_string(),
            java::JavaExtractor::create_parser,
            || Box::new(java::JavaExtractor),
        );

        registry.register(
            Php.name().to_string(),
            php::PhpExtractor::create_parser,
            || Box::new(php::PhpExtractor),
        );

        registry.register(
            CSharp.name().to_string(),
            csharp::CSharpExtractor::create_parser,
            || Box::new(csharp::CSharpExtractor),
        );

        registry.register(C.name().to_string(), c::CExtractor::create_parser, || {
            Box::new(c::CExtractor)
        });

        registry.register(
            Cpp.name().to_string(),
            cpp::CppExtractor::create_parser,
            || Box::new(cpp::CppExtractor),
        );

        registry.register(
            Haskell.name().to_string(),
            haskell::HaskellExtractor::create_parser,
            || Box::new(haskell::HaskellExtractor),
        );

        registry.register(
            Dart.name().to_string(),
            dart::DartExtractor::create_parser,
            || Box::new(dart::DartExtractor),
        );

        registry
    }

    fn register(
        &mut self,
        language: String,
        parser_factory: ParserFactory,
        extractor_factory: ExtractorFactory,
    ) {
        self.parsers.insert(language.clone(), parser_factory);
        self.extractors.insert(language, extractor_factory);
    }

    pub fn create_parser(&self, language: &str) -> Result<Parser> {
        self.parsers
            .get(language)
            .ok_or_else(|| {
                GitTypeError::ExtractionFailed(format!("Unsupported language: {}", language))
            })
            .and_then(|factory| factory())
    }

    pub fn get_extractor(&self, language: &str) -> Result<Box<dyn LanguageExtractor>> {
        self.extractors
            .get(language)
            .ok_or_else(|| {
                GitTypeError::ExtractionFailed(format!("Unsupported language: {}", language))
            })
            .map(|factory| factory())
    }

    pub fn create_query(&self, language: &str) -> Result<Query> {
        let extractor = self.get_extractor(language)?;
        let tree_sitter_lang = extractor.tree_sitter_language();
        let query_str = extractor.query_patterns();

        Query::new(&tree_sitter_lang, query_str).map_err(|e| {
            GitTypeError::ExtractionFailed(format!(
                "Failed to create query for {}: {}",
                language, e
            ))
        })
    }

    pub fn create_comment_query(&self, language: &str) -> Result<Query> {
        let extractor = self.get_extractor(language)?;
        let tree_sitter_lang = extractor.tree_sitter_language();
        let query_str = extractor.comment_query();

        Query::new(&tree_sitter_lang, query_str).map_err(|e| {
            GitTypeError::ExtractionFailed(format!(
                "Failed to create comment query for {}: {}",
                language, e
            ))
        })
    }

    pub fn supported_languages(&self) -> Vec<String> {
        self.parsers.keys().cloned().collect()
    }
}

static REGISTRY: Lazy<ParserRegistry> = Lazy::new(ParserRegistry::new);

pub fn get_parser_registry() -> &'static ParserRegistry {
    &REGISTRY
}

thread_local! {
    static TL_PARSERS: RefCell<std::collections::HashMap<String, Parser>> = RefCell::new(std::collections::HashMap::new());
}

/// Parse source using a thread-local parser per language to avoid re-allocations.
pub fn parse_with_thread_local(language: &str, content: &str) -> Option<Tree> {
    TL_PARSERS.with(|cell| {
        let mut map = cell.borrow_mut();
        let parser = match map.get_mut(language) {
            Some(p) => p,
            None => {
                // Create and insert parser if not exists
                match REGISTRY.create_parser(language) {
                    Ok(p) => {
                        map.insert(language.to_string(), p);
                        map.get_mut(language).unwrap()
                    }
                    Err(_) => return None,
                }
            }
        };
        parser.parse(content, None)
    })
}
