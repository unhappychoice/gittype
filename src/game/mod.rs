pub mod screens;
pub mod session;
pub mod text_processor;
pub mod display;
pub mod display_optimized;
pub mod display_ratatui;
pub mod challenge;
pub mod stage_manager;
pub mod stage_builder;

pub use screens::TypingScreen;
pub use session::GameSession;
pub use challenge::Challenge;
pub use stage_manager::StageManager;
pub use stage_builder::{StageBuilder, GameMode, DifficultyLevel, StageConfig};