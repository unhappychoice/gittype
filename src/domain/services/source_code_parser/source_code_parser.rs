use crate::domain::models::Language;
use crate::domain::models::{CodeChunk, ExtractionOptions};
use crate::domain::services::source_code_parser::parsers::parse_with_thread_local;
use crate::domain::services::source_code_parser::ChunkExtractor;
use crate::infrastructure::git::LocalGitRepositoryClient;
use crate::presentation::game::models::StepType;
use crate::presentation::game::screens::loading_screen::ProgressReporter;
use crate::{GitTypeError, Result};
use rayon::prelude::*;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::{
    atomic::{AtomicUsize, Ordering},
    Arc,
};

pub struct SourceCodeParser;

impl SourceCodeParser {
    pub fn new() -> Result<Self> {
        Ok(Self)
    }

    pub fn extract_chunks_with_progress<P: ProgressReporter + ?Sized>(
        &mut self,
        files_to_process: Vec<(PathBuf, Box<dyn Language>)>,
        options: &ExtractionOptions,
        progress: &P,
    ) -> Result<Vec<CodeChunk>> {
        let git_root = Self::find_git_root(&files_to_process)?;
        let valid_files = Self::filter_and_sort_files(files_to_process, options);
        let valid_files_count = valid_files.len();

        // Initialize extracting progress from 0
        let processed = Arc::new(AtomicUsize::new(0));
        progress.set_file_counts(StepType::Extracting, 0, valid_files_count, None);

        let all_chunks: Vec<CodeChunk> = valid_files
            .into_par_iter()
            .inspect(|_| {
                let current = processed.fetch_add(1, Ordering::Relaxed) + 1;
                Self::update_progress_if_needed(progress, current, valid_files_count);
            })
            .flat_map(|(path, language, _size)| {
                Self::read_and_parse_file(&git_root, &path, language).into_par_iter()
            })
            .flat_map(|(tree, content, file_path, git_root, language)| {
                ChunkExtractor::extract_chunks_from_tree(
                    &tree,
                    &content,
                    &file_path,
                    &git_root,
                    language.as_ref(),
                )
                .unwrap_or_default()
            })
            .collect();

        // Get final count and ensure final progress is exactly 100%
        let final_count = processed.load(Ordering::Relaxed);
        progress.set_file_counts(StepType::Extracting, final_count, final_count, None);
        progress.set_current_file(None);

        Ok(all_chunks)
    }

    fn find_git_root(files_to_process: &[(PathBuf, Box<dyn Language>)]) -> Result<PathBuf> {
        files_to_process
            .first()
            .map(|(first_file, _)| first_file)
            .and_then(|path| LocalGitRepositoryClient::get_repository_root(path))
            .ok_or_else(|| GitTypeError::ExtractionFailed("Git repository not found".to_string()))
    }

    fn filter_and_sort_files(
        files_to_process: Vec<(PathBuf, Box<dyn Language>)>,
        options: &ExtractionOptions,
    ) -> Vec<(PathBuf, Box<dyn Language>, u64)> {
        let mut valid_files: Vec<_> = files_to_process
            .into_par_iter()
            .filter_map(|(path, lang)| {
                let size = std::fs::metadata(&path).map(|m| m.len()).unwrap_or(0);
                if size > options.max_file_size_bytes {
                    log::warn!(
                        "Skipping large file: {:?} ({}MB > {}MB limit)",
                        path,
                        size / (1024 * 1024),
                        options.max_file_size_bytes / (1024 * 1024)
                    );
                    None
                } else {
                    Some((path, lang, size))
                }
            })
            .collect();

        valid_files.sort_by(|a, b| b.2.cmp(&a.2)); // Sort by size (largest first)
        valid_files
    }

    fn update_progress_if_needed<P: ProgressReporter + ?Sized>(
        progress: &P,
        current: usize,
        total_files: usize,
    ) {
        // Dynamic progress update frequency: every 10 files, but every 1 file when >99%
        let progress_threshold = if current as f64 / total_files as f64 > 0.99 {
            1 // Update every file when >99%
        } else {
            10 // Update every 10 files normally
        };

        if current.is_multiple_of(progress_threshold) || current == total_files {
            progress.set_file_counts(StepType::Extracting, current, total_files, None);
        }
    }

    #[allow(clippy::type_complexity)]
    fn read_and_parse_file(
        git_root: &Path,
        file_path: &Path,
        language: Box<dyn Language>,
    ) -> Option<(
        tree_sitter::Tree,
        String,
        PathBuf,
        PathBuf,
        Box<dyn Language>,
    )> {
        let content = fs::read_to_string(file_path).ok()?;
        let tree = parse_with_thread_local(language.name(), &content)?;

        Some((
            tree,
            content,
            file_path.to_path_buf(),
            git_root.to_path_buf(),
            language,
        ))
    }
}
