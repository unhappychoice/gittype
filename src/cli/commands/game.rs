use crate::cli::args::Cli;
use crate::extractor::ExtractionOptions;
use crate::game::screens::loading_screen::{LoadingScreen, ProcessingResult};
use crate::game::StageManager;
use crate::{GitTypeError, Result};
use std::path::PathBuf;

pub fn run_game_session(cli: Cli) -> Result<()> {
    let mut options = ExtractionOptions::default();

    if let Some(langs) = cli.langs {
        if let Err(unsupported_langs) =
            crate::extractor::models::language::LanguageRegistry::validate_languages(&langs)
        {
            eprintln!(
                "❌ Unsupported language(s): {}",
                unsupported_langs.join(", ")
            );
            eprintln!("💡 Supported languages:");
            let supported =
                crate::extractor::models::language::LanguageRegistry::get_supported_languages();
            let mut supported_display = supported.clone();
            supported_display.dedup();
            for chunk in supported_display.chunks(6) {
                eprintln!("   {}", chunk.join(", "));
            }
            std::process::exit(1);
        }

        options.languages = Some(langs);
        options.apply_language_filter();
    }

    let repo_spec = cli.repo.as_deref();
    let default_repo_path = cli.repo_path.unwrap_or_else(|| PathBuf::from("."));
    let initial_repo_path = if repo_spec.is_some() {
        None
    } else {
        Some(&default_repo_path)
    };

    let session_result = LoadingScreen::new()
        .and_then(|mut loading_screen| {
            let result = loading_screen.process_repository(repo_spec, initial_repo_path, &options);
            let _ = loading_screen.cleanup();
            result
        })
        .and_then(|result| {
            if result.challenges.is_empty() {
                Err(GitTypeError::NoSupportedFiles)
            } else {
                Ok(result)
            }
        })
        .and_then(
            |ProcessingResult {
                 challenges,
                 git_info,
             }| {
                let mut stage_manager = StageManager::new(challenges);
                stage_manager.set_git_info(git_info);
                stage_manager.run_session()
            },
        );

    match session_result {
        Ok(_) => {}
        Err(e) => handle_game_error(e)?,
    }

    Ok(())
}

fn handle_game_error(e: GitTypeError) -> Result<()> {
    match e {
        GitTypeError::NoSupportedFiles => {
            eprintln!("❌ No code chunks found in the repository");
            eprintln!("💡 Try:");
            eprintln!("   • Using a different repository path");
            eprintln!("   • Adjusting --langs filter (e.g., --langs rust,python)");
            std::process::exit(1);
        }
        GitTypeError::RepositoryNotFound(path) => {
            eprintln!("❌ Repository not found at path: {}", path.display());
            eprintln!("💡 Ensure the path exists and is a valid repository");
            std::process::exit(1);
        }
        GitTypeError::RepositoryCloneError(git_error) => {
            eprintln!("❌ Failed to clone repository: {}", git_error);
            eprintln!("💡 Check:");
            eprintln!("   • Repository URL is correct");
            eprintln!("   • You have access to the repository");
            eprintln!("   • Internet connection is available");
            std::process::exit(1);
        }
        GitTypeError::ExtractionFailed(msg) => {
            eprintln!("❌ Code extraction failed: {}", msg);
            eprintln!("💡 Try using different --langs filter");
            std::process::exit(1);
        }
        GitTypeError::InvalidRepositoryFormat(msg) => {
            eprintln!("❌ Invalid repository format: {}", msg);
            eprintln!("💡 Supported formats:");
            eprintln!("   • owner/repo");
            eprintln!("   • https://github.com/owner/repo");
            eprintln!("   • git@github.com:owner/repo.git");
            std::process::exit(1);
        }
        GitTypeError::IoError(io_error) => {
            eprintln!("❌ IO error: {}", io_error);
            std::process::exit(1);
        }
        GitTypeError::DatabaseError(db_error) => {
            eprintln!("❌ Database error: {}", db_error);
            std::process::exit(1);
        }
        GitTypeError::GlobPatternError(glob_error) => {
            eprintln!("❌ Invalid glob pattern: {}", glob_error);
            eprintln!("💡 Check your glob patterns in ExtractionOptions");
            std::process::exit(1);
        }
        GitTypeError::SerializationError(json_error) => {
            eprintln!("❌ Serialization error: {}", json_error);
            std::process::exit(1);
        }
        GitTypeError::TerminalError(msg) => {
            eprintln!("❌ Terminal error: {}", msg);
            if msg.contains("No such device or address") {
                eprintln!("💡 This error often occurs in WSL or SSH environments where terminal features are limited.");
                eprintln!("   Try running GitType in a native terminal or GUI terminal emulator.");
            }
            std::process::exit(1);
        }
        GitTypeError::WalkDirError(walk_error) => {
            eprintln!("❌ Directory walk error: {}", walk_error);
            eprintln!("💡 Check directory permissions and try again");
            std::process::exit(1);
        }
        GitTypeError::TreeSitterLanguageError(lang_error) => {
            eprintln!("❌ Language parsing error: {}", lang_error);
            eprintln!("💡 This might be caused by unsupported language features");
            std::process::exit(1);
        }
    }
}
