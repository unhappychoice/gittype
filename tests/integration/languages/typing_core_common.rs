use gittype::domain::services::source_code_parser::parsers::parse_with_thread_local;
use gittype::domain::services::source_code_parser::CommonExtractor;
use gittype::presentation::game::typing_core::{ProcessingOptions, TypingCore};
use insta::assert_snapshot;
use std::fmt;

#[derive(Debug)]
pub struct TypingSnapshot {
    pub text_original: String,
    pub text_to_type: String,
    pub text_to_display: String,
}

impl fmt::Display for TypingSnapshot {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "# text_original\n{}\n\n# text_to_type\n{}\n\n# text_to_display\n{}\n",
            self.text_original, self.text_to_type, self.text_to_display
        )
    }
}

/// Test case for typing core functionality
pub struct TypingCoreTestCase {
    pub name: &'static str,
    pub code: &'static str,
    pub comment_ranges: Vec<(usize, usize)>,
    pub options: Option<ProcessingOptions>,
}

/// Common function to verify typing simulation works correctly
pub fn verify_typing_simulation(core: &mut TypingCore) {
    let mut chars_typed = Vec::new();

    while let Some(current_char_type) = core.current_char_to_type() {
        let current_char_display = core
            .text_to_display()
            .chars()
            .nth(core.current_position_to_display());

        // Test display position mapping
        let display_pos = core.current_position_to_display();
        println!(
            "Position {}: type_char='{}', display_char={:?}, display_pos={:?}",
            core.current_position_to_type(),
            current_char_type,
            current_char_display,
            display_pos
        );

        // Verify current_char_type == current_char_display (with special handling for newlines)
        match current_char_display {
            Some(display_char) => {
                if current_char_type == '\n' && display_char == '↵' {
                    // This is expected: newlines in type text map to ↵ symbols in display text
                    println!("  -> Newline mapping OK: '\\n' -> '↵'");
                } else if current_char_type == '\t' && display_char == '→' {
                    // This is expected: tabs in type text map to → symbols in display text
                    println!("  -> Tab mapping OK: '\\t' -> '→'");
                } else if current_char_type != display_char {
                    panic!(
                        "Mismatch at position {}: type_char='{}' != display_char='{}'",
                        core.current_position_to_type(),
                        current_char_type,
                        display_char
                    );
                }
            }
            None => {
                panic!(
                    "No display character at position {}",
                    core.current_position_to_type()
                );
            }
        }

        chars_typed.push(current_char_type);
        core.advance_to_next_character();
    }

    println!("Total characters typed: {}", chars_typed.len());
}

/// Create a snapshot for typing core test
pub fn create_typing_snapshot(core: &TypingCore, original_text: &str) -> TypingSnapshot {
    TypingSnapshot {
        text_original: original_text.to_string(),
        text_to_type: core.text_to_type().to_string(),
        text_to_display: core.text_to_display().to_string(),
    }
}

/// Run a complete typing core test
pub fn run_typing_core_test(test_case: TypingCoreTestCase) {
    let options = test_case.options.unwrap_or_default();
    let mut core = TypingCore::new(test_case.code, &test_case.comment_ranges, options);

    // Verify typing simulation works correctly
    verify_typing_simulation(&mut core);

    let snapshot = create_typing_snapshot(&core, test_case.code);
    assert_snapshot!(test_case.name, snapshot);
}

/// Find comment ranges using the same parser as CodeChunk extraction
pub fn find_comment_ranges_with_parser(code: &str, language: &str) -> Vec<(usize, usize)> {
    if let Some(tree) = parse_with_thread_local(language, code) {
        CommonExtractor::extract_comment_ranges(&tree, code, language, &[]).unwrap_or_default()
    } else {
        Vec::new()
    }
}

/// Macro to define typing core test cases using the real parser
#[macro_export]
macro_rules! typing_core_test_with_parser {
    ($test_name:ident, $lang:expr, $code:expr) => {
        #[test]
        fn $test_name() {
            use $crate::integration::languages::typing_core_common::*;

            let comment_ranges = find_comment_ranges_with_parser($code, $lang);
            let test_case = TypingCoreTestCase {
                name: stringify!($test_name),
                code: $code,
                comment_ranges,
                options: None,
            };

            run_typing_core_test(test_case);
        }
    };

    ($test_name:ident, $lang:expr, $code:expr, $options:expr) => {
        #[test]
        fn $test_name() {
            use $crate::integration::languages::typing_core_common::*;

            let comment_ranges = find_comment_ranges_with_parser($code, $lang);
            let test_case = TypingCoreTestCase {
                name: stringify!($test_name),
                code: $code,
                comment_ranges,
                options: Some($options),
            };

            run_typing_core_test(test_case);
        }
    };
}
