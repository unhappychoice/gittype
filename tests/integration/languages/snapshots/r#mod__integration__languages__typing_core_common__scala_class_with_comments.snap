---
source: tests/integration/languages/typing_core_common.rs
expression: snapshot
---
# text_original
class Person(val name: String, val age: Int) {
    // Person's greeting method
    def greet(): String = s"Hello, I'm $name"
    
    /* Check if person is adult
       Returns true if age >= 18 */
    def isAdult: Boolean = age >= 18
}

# text_to_type
class Person(val name: String, val age: Int) {
// Person's greeting method
def greet(): String = s"Hello, I'm $name"
/* Check if person is adult
Returns true if age >= 18 */
def isAdult: Boolean = age >= 18
}

# text_to_display
class Person(val name: String, val age: Int) {↵
    // Person's greeting method↵
    def greet(): String = s"Hello, I'm $name"↵
    
    /* Check if person is adult↵
       Returns true if age >= 18 */↵
    def isAdult: Boolean = age >= 18↵
}↵
